kernel/declarations.mli

```diff
@@ -287,6 +288,44 @@ type mutual_inductive_body = {
 
 type mind_specif = mutual_inductive_body * one_inductive_body
 
+(** {6 Rewrite rules } *)
+
+type instance_mask = bool array option
+
+type 'arg head_pattern =
+  | PHRel     of int
+  | PHSort    of Sorts.family
+  | PHSymbol  of Constant.t * instance_mask
+  | PHInd     of inductive * instance_mask
+  | PHConstr  of constructor * instance_mask
+  | PHInt     of Uint63.t
+  | PHFloat   of Float64.t
+  | PHLambda  of 'arg array * 'arg
+  | PHProd    of 'arg array * 'arg
+
+type pattern_elimination =
+  | PEApp     of pattern_argument array
+  | PECase    of inductive * instance_mask * pattern_argument * pattern_argument array
+  | PEProj    of Projection.t
+
+and head_elimination = pattern_argument head_pattern * pattern_elimination list
+
+and pattern_argument =
+  | EHole
+  | EHoleIgnored
+  | ERigid of head_elimination
+
+type rewrite_rule = {
+  lhs_pat : instance_mask * pattern_elimination list;
+  rhs : constr;
+}
+
+(** {6 Representation of rewrite rules in the kernel } *)
+
+type rewrite_rules_body = {
+  rewrules_rules : (Constant.t * rewrite_rule) list;
```

Comment:
More like `ps[c@{u]`, since eliminations may be other things than applications.
You're right about universes, and terms are matched in the same ascending order, so `Rel 1` is the first hole found.

This design is good in that patterns are linear by construction, but I don't know whether this property is really useful here or if it should be restricted to MetaCoq.

Reply:
>This design is good in that patterns are linear by construction

Isn't that something we may want to change someday?

Regardless I don't think it's enough of a benefit to justify picking this design IMO.